File: src/main/python/AoC2024_06.py
#! /usr/bin/env python3
#
# Advent of Code 2024 Day 6
#

from __future__ import annotations

import sys
from collections import OrderedDict
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Iterator

from typing import NamedTuple

from aoc.common import InputData
from aoc.common import SolutionBase
from aoc.common import aoc_samples
from aoc.grid import CharGrid

Cell = tuple[int, int]
Direction = tuple[int, int]
Input = CharGrid
Output1 = int
Output2 = int

UP, RIGHT, DOWN, LEFT = (0, 1), (1, 0), (0, -1), (-1, 0)
DIRS = [UP, RIGHT, DOWN, LEFT]


TEST = """\
....#.....
.........#
..........
..#.......
.......#..
..........
.#..^.....
........#.
#.........
......#...
"""


class Obstacles(NamedTuple):
    obs: dict[Direction, list[list[Cell | None]]]

    @classmethod
    def _obstacles(
        cls, grid: CharGrid, starts: Iterator[Cell], direction: Direction
    ) -> list[list[Cell | None]]:
        h, w = grid.get_height(), grid.get_width()
        obs: list[list[Cell | None]] = [
            [None for _ in range(w)] for _ in range(h)
        ]
        for start in starts:
            last = start if grid.values[start[0]][start[1]] == "#" else None
            cell = start
            while True:
                r, c = cell
                obs[r][c] = last
                if grid.values[r][c] == "#":
                    last = cell
                nr, nc = r - direction[1], c + direction[0]
                if not (0 <= nr < h and 0 <= nc < w):
                    break
                cell = (nr, nc)
        return obs

    @classmethod
    def from_grid(cls, grid: CharGrid) -> Obstacles:
        h, w = grid.get_height(), grid.get_width()
        obs = dict[Direction, list[list[Cell | None]]]()
        mci, mri = grid.get_max_col_index(), grid.get_max_row_index()
        obs[RIGHT] = Obstacles._obstacles(
            grid, ((r, mci) for r in range(h)), LEFT
        )
        obs[LEFT] = Obstacles._obstacles(
            grid, ((r, 0) for r in range(h)), RIGHT
        )
        obs[UP] = Obstacles._obstacles(grid, ((0, c) for c in range(w)), DOWN)
        obs[DOWN] = Obstacles._obstacles(
            grid, ((mri, c) for c in range(w)), UP
        )
        return Obstacles(obs)

    def get_next(
        self, start: Cell, direction: Direction, extra: Cell
    ) -> Cell | None:
        obs = self.obs[direction][start[0]][start[1]]
        if obs is None:
            if (
                direction[1] == 0
                and start[0] == extra[0]
                and direction == (RIGHT if start[1] < extra[1] else LEFT)
            ) or (
                direction[0] == 0
                and start[1] == extra[1]
                and direction == (DOWN if start[0] < extra[0] else UP)
            ):
                return extra
            return None

        if (
            direction[1] == 0
            and obs[0] == extra[0]
            and direction == (RIGHT if start[1] < extra[1] else LEFT)
        ):
            dextra = abs(extra[1] - start[1])
            dobs = abs(obs[1] - start[1])
            return obs if dobs < dextra else extra

        if (
            direction[0] == 0
            and obs[1] == extra[1]
            and direction == (DOWN if start[0] < extra[0] else UP)
        ):
            dextra = abs(extra[0] - start[0])
            dobs = abs(obs[0] - start[0])
            return obs if dobs < dextra else extra

        return obs


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return CharGrid.from_strings(list(input_data))

    def visited(
        self,
        grid: CharGrid,
        direction: int,
    ) -> OrderedDict[Cell, list[int]]:
        start = next(grid.get_all_equal_to("^"))
        pos = (start.row, start.col)
        h, w = grid.get_height(), grid.get_width()
        seen = OrderedDict[Cell, list[int]]()
        seen.setdefault(pos, []).append(direction)
        while True:
            nxt = (pos[0] - DIRS[direction][1], pos[1] + DIRS[direction][0])
            if not (0 <= nxt[0] < h and 0 <= nxt[1] < w):
                return seen
            if grid.values[nxt[0]][nxt[1]] == "#":
                direction = (direction + 1) % len(DIRS)
            else:
                pos = nxt
            seen.setdefault(pos, []).append(direction)

    def is_loop(
        self, pos: Cell, direction: int, obs: Obstacles, extra: Cell
    ) -> bool:
        seen = dict[Cell, int]()
        seen[pos] = 1 << direction
        while True:
            nxt_obs = obs.get_next(pos, DIRS[direction], extra)
            if nxt_obs is None:
                return False
            pos = (
                nxt_obs[0] + DIRS[direction][1],
                nxt_obs[1] - DIRS[direction][0],
            )
            direction = (direction + 1) % len(DIRS)
            if pos in seen and seen[pos] & (1 << direction) != 0:
                return True
            seen[pos] = seen.get(pos, 0) | (1 << direction)

    def part_1(self, grid: Input) -> Output1:
        return len(self.visited(grid, 0))

    def part_2(self, grid: Input) -> Output2:
        obs = Obstacles.from_grid(grid)
        visited = self.visited(grid, 0)
        prev_pos, prev_dirs = visited.popitem(last=False)
        ans = 0
        while len(visited) > 0:
            pos, dirs = visited.popitem(last=False)
            ans += self.is_loop(prev_pos, prev_dirs.pop(0), obs, pos)
            prev_pos, prev_dirs = pos, dirs
        return ans

    @aoc_samples(
        (
            ("part_1", TEST, 41),
            ("part_2", TEST, 6),
        )
    )
    def samples(self) -> None:
        pass


solution = Solution(2024, 6)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2024_06.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Iterator;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.Map.Entry;
import java.util.Optional;
import java.util.function.ToIntFunction;
import java.util.stream.Stream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.Grid.Cell;
import com.github.pareronia.aoc.geometry.Direction;
import com.github.pareronia.aoc.geometry.Turn;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

public final class AoC2024_06
        extends SolutionBase<CharGrid, Integer, Integer> {
    
    private AoC2024_06(final boolean debug) {
        super(debug);
    }
    
    public static AoC2024_06 create() {
        return new AoC2024_06(false);
    }
    
    public static AoC2024_06 createDebug() {
        return new AoC2024_06(true);
    }
    
    @Override
    protected CharGrid parseInput(final List<String> inputs) {
        return new CharGrid(inputs);
    }
    
    private LinkedHashMap<Cell, List<Direction>> visited(
        final CharGrid grid, Direction dir
    ) {
        Cell pos = grid.getAllEqualTo('^').findFirst().orElseThrow();
        final LinkedHashMap<Cell, List<Direction>> visited
            = new LinkedHashMap<>(Map.of(pos, new ArrayList<>(List.of(dir))));
        while (true) {
            final Cell nxt = pos.at(dir);
            if (!grid.isInBounds(nxt)) {
                return visited;
            }
            if (grid.getValue(nxt) == '#') {
                dir = dir.turn(Turn.RIGHT);
            } else {
                pos = nxt;
            }
            visited.computeIfAbsent(pos, k -> new ArrayList<>()).add(dir);
        }
    }
    
    @Override
    public Integer solvePart1(final CharGrid grid) {
        return visited(grid, Direction.UP).size();
    }
    
    private boolean isLoop(
            Cell pos, Direction dir, final Obstacles obs, final Cell extra
    ) {
        final ToIntFunction<Direction> bits = d -> switch (d) {
            case UP -> 1;
            case RIGHT -> 1 << 1;
            case DOWN -> 1 << 2;
            case LEFT -> 1 << 3;
            default -> throw new IllegalArgumentException();
        };
        final Map<Cell, Integer> seen = new HashMap<>();
        seen.put(pos, bits.applyAsInt(dir));
        while (true) {
            final Optional<Cell> nextObs = obs.getNext(pos, dir, extra);
            if (nextObs.isEmpty()) {
                return false;
            }
            pos = nextObs.get().at(dir.turn(Turn.AROUND));
            dir = dir.turn(Turn.RIGHT);
            final int bDir = bits.applyAsInt(dir);
            if (seen.containsKey(pos) && (seen.get(pos) & bDir) != 0) {
                return true;
            }
            seen.merge(pos, bDir, (a, b) -> a |= b);
        }
    }
    
    @Override
    public Integer solvePart2(final CharGrid grid) {
        final Obstacles obs = Obstacles.from(grid);
        final LinkedHashMap<Cell, List<Direction>> visited
                = visited(grid, Direction.UP);
        final Iterator<Entry<Cell, List<Direction>>> it
                = visited.entrySet().iterator();
        Entry<Cell, List<Direction>> prev = it.next();
        int ans = 0;
        while (it.hasNext()) {
            final Entry<Cell, List<Direction>> curr = it.next();
            final Cell start = prev.getKey();
            final Direction dir = prev.getValue().remove(0);
            if (isLoop(start, dir, obs, curr.getKey())) {
                ans += 1;
            }
            prev = curr;
        }
        return ans;
    }
    
    @Override
    @Samples({
        @Sample(method = "part1", input = TEST, expected = "41"),
        @Sample(method = "part2", input = TEST, expected = "6"),
    })
    public void samples() {
    }
    
    public static void main(final String[] args) throws Exception {
        AoC2024_06.create().run();
    }

    private static final String TEST = """
            ....#.....
            .........#
            ..........
            ..#.......
            .......#..
            ..........
            .#..^.....
            ........#.
            #.........
            ......#...
            """;

    record Obstacles(Map<Direction, Cell[][]> obs) {
        private static Cell[][] obstacles(
                final CharGrid grid,
                final Stream<Cell> starts,
                final Direction dir
        ) {
            final Cell[][] obs = new Cell[grid.getHeight()][grid.getWidth()];
            starts.forEach(start -> {
                Cell last = grid.getValue(start) == '#' ? start : null;
                final Iterator<Cell> it
                        = grid.getCells(start, dir).iterator();
                while (it.hasNext()) {
                    final Cell cell = it.next();
                    obs[cell.getRow()][cell.getCol()] = last;
                    if (grid.getValue(cell) == '#') {
                        last = cell;
                    }
                }
            });
            return obs;
        }

        public static Obstacles from(final CharGrid grid) {
            final Map<Direction, Cell[][]> obs = new HashMap<>();
            obs.put(Direction.RIGHT, Obstacles.obstacles(
                    grid,
                    range(grid.getHeight()).intStream()
                        .mapToObj(r -> Cell.at(r, grid.getMaxColIndex())),
                    Direction.LEFT));
            obs.put(Direction.LEFT, Obstacles.obstacles(
                    grid,
                    range(grid.getHeight()).intStream()
                        .mapToObj(r -> Cell.at(r, 0)),
                    Direction.RIGHT));
            obs.put(Direction.UP, Obstacles.obstacles(
                    grid,
                    range(grid.getWidth()).intStream()
                        .mapToObj(c -> Cell.at(0, c)),
                    Direction.DOWN));
            obs.put(Direction.DOWN, Obstacles.obstacles(
                    grid,
                    range(grid.getWidth()).intStream()
                        .mapToObj(c -> Cell.at(grid.getMaxRowIndex(), c)),
                    Direction.UP));
            return new Obstacles(obs);
        }

        public Optional<Cell> getNext(
                final Cell start, final Direction dir, final Cell extra
        ) {
            final Cell obs = this.obs.get(dir)[start.getRow()][start.getCol()];
            if (obs == null) {
                if (dir.isHorizontal()
                        && start.getRow() == extra.getRow()
                        && start.to(extra) == dir
                ) {
                    return Optional.of(extra);
                } else if (dir.isVertical()
                        && start.getCol() == extra.getCol()
                        && start.to(extra) == dir
                ) {
                    return Optional.of(extra);
                }
                return Optional.empty();
            } else {
                int dExtra = 0;
                int dObs = 0;
                if (dir.isHorizontal()
                        && obs.getRow() == extra.getRow()
                        && start.to(extra) == dir
                ) {
                    dExtra = Math.abs(extra.getCol() - start.getCol());
                    dObs = Math.abs(obs.getCol() - start.getCol());
                    return Optional.of(dObs < dExtra ? obs : extra);
                } else if (dir.isVertical()
                        && obs.getCol() == extra.getCol()
                        && start.to(extra) == dir
                ) {
                    dExtra = Math.abs(extra.getRow() - start.getRow());
                    dObs = Math.abs(obs.getRow() - start.getRow());
                    return Optional.of(dObs < dExtra ? obs : extra);
                }
                return Optional.of(obs);
            }
        }
    }
}